libs/datagrid/src/lib/ag-grid/ag-grid-lazy-themed.tsx

```diff
@@ -1,30 +1,78 @@
+import type { MutableRefObject } from 'react';
+import React, { useCallback, useRef } from 'react';
 import type { AgGridReactProps, AgReactUiProps } from 'ag-grid-react';
+import type { ColumnResizedEvent, GridReadyEvent } from 'ag-grid-community';
 import { AgGridReact } from 'ag-grid-react';
-import { useThemeSwitcher } from '@vegaprotocol/react-helpers';
+import {
+  useResizeObserver,
+  useThemeSwitcher,
+} from '@vegaprotocol/react-helpers';
+import { useColumnSizes } from './use-column-sizes';
 import classNames from 'classnames';
 
 export const AgGridThemed = ({
   style,
   gridRef,
+  storeKey,
   ...props
 }: (AgGridReactProps | AgReactUiProps) & {
   style?: React.CSSProperties;
   gridRef?: React.ForwardedRef<AgGridReact>;
+  storeKey?: string;
 }) => {
+  const containerRef = useRef<HTMLDivElement | null>(null);
+  const [sizes, setValues, onResizeCallback] = useColumnSizes({
+    storeKey,
+    container: containerRef,
+  });
   const { theme } = useThemeSwitcher();
   const defaultProps = {
     rowHeight: 22,
     headerHeight: 22,
     enableCellTextSelection: true,
+    onColumnResized: useCallback(
+      (event: ColumnResizedEvent) => {
+        if (event.source === 'uiColumnDragged' && event.columns) {
+          setValues(event.columns);
+        }
+      },
+      [setValues]
+    ),
   };
+  const { onGridReady } = props;
+  const onGridReadyInternal = useCallback(
+    (event: GridReadyEvent) => {
+      onGridReady?.(event);
```

Comment:
@macqbat I don't like all this logic being in this component. What if you want to just render a simple grid that doesn't store column widths or one that just uses a flex layout? For example in the governance app. They now have local storage entries which are not required.

I would suggest moving more of the logic to the hook and let the consumer dictate when to use the store.

If we can get the hook's API to look something more like the below it would be more explicit and declarative and doesn't require you to know the inner workings of our AgGrid component (IE what happens if you use storeKey). As well as not polluting other apps localstorage.

This also has the added benefit of tidying hook logic because we can guarantee that the required arguments are provided.

```
const MyGrid = forwardRef((props, ref) => {
  const { onGridReady, onColumnResized } = useColumnSizes({
    storeKey: 'MyGrid',
    gridRef: ref
  })

  return (
    <AgGird
       ref={ref}
       onGridReady={onGridReady}
       onColumnResized={onColumnResized}
    />
  )
})
```

If you really dont want to have to useColumnSizes in multiple places I would suggest an HOC around the grid component

```
const PersistedDataGrid = ({ storeKey: { storeKey: string }) => {
    const callbacks = useColumnSizes({
      storeKey,
      ...
    });
    return <AgGrid {...callbacks} />
}
```

Reply:
I'm afraid the use of HOC here would't reduce amount of places where changes should be implemented, since our usual hierarchy of ag-grid components (container => manager => table) is quite deep, and potential HOC should somehow passing it all by as well as take into consideration lazy import of ag-grid children.
But I made some tidying up as you've suggested, as closing up more logic into hook, and in case of not adding storeKey all other components can actually omit results of this hook. I think it looks quite good now.. :-)